examples/win-offsets.c

```diff
@@ -51,57 +78,164 @@ void sigint_handler()
     exit(1);
 }
 
+event_response_t cr3_cb(vmi_instance_t vmi, vmi_event_t *event)
+{
+    dp("CR#: %llx, VCPU: %d\n", event->reg_event.value, event->vcpu_id);
+
+    if (event->vcpu_id != 0) {
+        /* LibVMI initialization procedure strongly
+         * relies on reading CPU context for VCPU 0
+         * so we only do care about this one. */
+        return VMI_EVENT_RESPONSE_NONE;
+    }
+
+    addr_t gs_base;
+    addr_t kthread;
+    addr_t eprocess;
+
+    access_context_t ctx = {
+        .translate_mechanism = VMI_TM_PROCESS_DTB,
+        .dtb = event->reg_event.value
+    };
+
+    /*
+     * Get current GS_BASE and translate its VA to PA using current CR3.
+     * This may fail (most probably) if we are in user-mode DTB with KPTI hardening.
+     */
+    page_mode_t pm = vmi_get_page_mode(vmi, 0);
+    reg_t kpcr_reg = pm == VMI_PM_IA32E ? GS_BASE : FS_BASE;
+
+    if (VMI_FAILURE == vmi_get_vcpureg(vmi, &gs_base, kpcr_reg, 0)) {
```

Comment:
Why not just use the registers that are passed down as part of the event?

Reply:
fixed